src/pkg/golang/build.go

```diff
@@ -204,19 +232,107 @@ func (c Environ) GoCmd(gocmd string, args ...string) *exec.Cmd {
 	return cmd
 }
 
-// Version returns the Go version string that runtime.Version would return for
-// the Go compiler in this environ.
-func (c Environ) Version() (string, error) {
+// resolve absolute path of go-compiler option from PATH
+func (c *Environ) resolveCompiler() error {
+	if c.Compiler.Path != "" {
+		fname, err := exec.LookPath(string(c.Compiler.Path))
+		if err == nil {
+			fname, err = filepath.Abs(fname)
+		}
+		if err != nil {
+			return fmt.Errorf("build: %v", err)
+		}
+		c.Compiler.Path = fname
+	}
+	return nil
+}
+
+// runs GoCmd("version") and parse/caches output, to environ.Compiler
+func (c *Environ) InitCompiler() error {
+
+	if c.Compiler.IsInit {
+		return nil
+	}
+
+	c.resolveCompiler()
+
 	cmd := c.GoCmd("version")
 	v, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", err
+		return err
 	}
+
+	efmt := "go-compiler 'version' output unrecognized: %v"
 	s := strings.Fields(string(v))
-	if len(s) < 3 {
-		return "", fmt.Errorf("unknown go version, tool returned weird output for 'go version': %v", string(v))
+	if len(s) < 1 {
+		return fmt.Errorf(efmt, string(v))
 	}
-	return s[2], nil
+
+	compiler := c.Compiler
+	compiler.Identifier = s[0]
+	compiler.VersionOutput = string(v)
+	compiler.IsInit = true
+
+	switch compiler.Identifier {
+
+	case "go":
```

Comment:
What do you think of refactoring this into smaller chunks and make it a bit more generic, such that the it becomes a bit more readable?

```go
type CompilerType int
const (
	CompilerGo CompilerType = iota
	CompilerTinygo
	CompilerUnkown
)

func CompilerTypeFromString(name string) CompilerType {
	val, ok := map[string]CompilerType{
		"go":     CompilerGo,
		"tinygo": CompilerTinygo,
	}[name]
	if ok {
		return val
	}
	return CompilerUnkown
}
```

Further one could implement a parsing function over the Compiler Type so that the parsing logic is put there.

```go
func (c *Compiler) Parse(cType CompilerType) (err error){
  switch cType {
    case CompilerGo:
      // parse
    
    case CompilerTinygo:
      // parse
    
    case CompilerUnkown:
      // parse
    
    default:
      return fmt.Errorf("unkown")
  }
}
```

```go
compilerID := CompilerTypeFromString(compiler.Identifier)
if err := compiler.Parse(compilerID); err != nil {
    // handle error
}
```

Reply:
I would be happy to! Thanks for the feedback!

[I'm just beginning to learn go]

Reply:
+1 on this